cesql/spec.md

```diff
@@ -210,9 +225,24 @@ returned together with the evaluated value of the CESQL expression.
 Whenever possible, some error checks SHOULD be done at compile time by the expression evaluator, in order to prevent
 runtime errors.
 
+Every CESQL engine MUST support the following error types:
+
+- _ParseError_: An error that occurs during parseing
+- _MathError_: An error that occurs during the evaluation of a mathematical operation
+- _CastError_: An error that occurs during an implicit or explicit type cast
+- _MissingFunctionError_: An error that occurs due to a call to a function that has not been registered with the CESQL engine
+- _FunctionEvaluationError_: An error that occurs during the evaluation of a function
+
+Whenever an operator or function encounters an error, it MUST return a return value as well as an error value or values. In cases where there is not an obvious return value for the expression,
```

Comment:
See if this is easier to read:
...it MUST result in a "return value" as well as one or more "error values".

Up to you

Reply:
Is a missing attribute an error? Kind of feel like it should be and therefore should have an error type, no? I could be swayed into believing that it's special, but if so we may want to say that a missing attribute isn't an error in the spec so people don't think we overlooked it.

Also, I can't help but wonder if there should be some kind of "generic error" define for cases that you didn't think of above.

Reply:
I keep going back to whether or not a missing attribute should be an error. I think with the introduction of zero values for each type along with the short circuiting of AND/OR, it makes sense to have missing attributes as an error again. A big part of the argument for not having an error was to try and understand an expression like:
```sql
missingAttribute AND true
```
as well as
```
true OR missingAttribute
```

But, with these new changes we have made I think that these can be reasoned about so including the error (with the zero type) makes more sense